File: lib/features/imports/applehealth/apple_health_import_record_conversions.dart
part of 'apple_health_import_converter.dart';

/// Body / vitals / hydration / cycle / mindfulness single-record conversions,
/// ported from the Kotlin `AppleHealthImportRecordConversions.kt`.
extension AppleHealthImportRecordConversions on AppleHealthImportConverter {
  ConvertedAppleRecord _instantRecord(
    AppleRecord record,
    AppleDateTime start,
    String targetType,
    String fingerprintPrefix,
    ImportRecord Function(String clientRecordId) build, {
    String? valueOverride,
  }) {
    final fingerprint = record.stableClientRecordId(fingerprintPrefix);
    markConverted(record.type);
    return ConvertedAppleRecord(
      appleType: record.type,
      targetType: targetType,
      fingerprint: fingerprint,
      record: build(_metadataClientRecordId(record, targetType)),
      sourceTimeRange: AppleImportTimeRange(start.instant, start.instant),
      unit: record.unit,
      value: valueOverride ?? record.valueForReport,
    );
  }

  ConvertedAppleRecord? convertWeight(AppleRecord record, AppleDateTime start) {
    final value = record.numericValue;
    final kg = value == null ? null : toKilograms(value, record.unit);
    if (kg == null || kg <= 0.0) {
      return invalidRecord(
        record,
        'Weight is missing, unsupported unit, or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'WeightRecord',
      'weight',
      (clientRecordId) => WeightImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        kilograms: kg,
      ),
    );
  }

  ConvertedAppleRecord? convertHeight(AppleRecord record, AppleDateTime start) {
    final value = record.numericValue;
    final meters = value == null ? null : toMeters(value, record.unit);
    if (meters == null || meters <= 0.0) {
      return invalidRecord(
        record,
        'Height is missing, unsupported unit, or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'HeightRecord',
      'height',
      (clientRecordId) => HeightImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        meters: meters,
      ),
    );
  }

  ConvertedAppleRecord? convertBodyFat(AppleRecord record, AppleDateTime start) {
    final value = record.numericValue;
    final percent = value == null ? null : toPercentage(value, record.unit);
    if (percent == null || percent < 0.0 || percent > 100.0) {
      return invalidRecord(
        record,
        'Body fat is missing, unsupported unit, or outside 0..100%.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BodyFatRecord',
      'body_fat',
      (clientRecordId) => BodyFatImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        percent: percent,
      ),
    );
  }

  ConvertedAppleRecord? convertLeanMass(AppleRecord record, AppleDateTime start) {
    final value = record.numericValue;
    final kg = value == null ? null : toKilograms(value, record.unit);
    if (kg == null || kg <= 0.0) {
      return invalidRecord(
        record,
        'Lean body mass is missing, unsupported unit, or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'LeanBodyMassRecord',
      'lean_mass',
      (clientRecordId) => LeanBodyMassImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        kilograms: kg,
      ),
    );
  }

  ConvertedAppleRecord? convertBoneMass(AppleRecord record, AppleDateTime start) {
    final value = record.numericValue;
    final kg = value == null ? null : toKilograms(value, record.unit);
    if (kg == null || kg <= 0.0) {
      return invalidRecord(
        record,
        'Bone mass is missing, unsupported unit, or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BoneMassRecord',
      'bone_mass',
      (clientRecordId) => BoneMassImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        kilograms: kg,
      ),
    );
  }

  ConvertedAppleRecord? convertBodyWaterMass(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final value = record.numericValue;
    final kg = value == null ? null : toKilograms(value, record.unit);
    if (kg == null || kg <= 0.0) {
      return invalidRecord(
        record,
        'Body water mass is missing, unsupported unit, or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BodyWaterMassRecord',
      'body_water_mass',
      (clientRecordId) => BodyWaterMassImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        kilograms: kg,
      ),
    );
  }

  ConvertedAppleRecord? convertHydration(AppleRecord record, AppleInterval iv) {
    final value = record.numericValue;
    final milliliters = value == null ? null : toMilliliters(value, record.unit);
    if (milliliters == null || milliliters <= 0.0) {
      return invalidRecord(
        record,
        'Hydration is missing, unsupported unit, or not positive.',
      );
    }
    final fingerprint = record.stableClientRecordId('hydration');
    markConverted(record.type);
    return ConvertedAppleRecord(
      appleType: record.type,
      targetType: 'HydrationRecord',
      fingerprint: fingerprint,
      record: HydrationImportRecord(
        clientRecordId: _metadataClientRecordId(record, 'HydrationRecord'),
        startTime: iv.start.instant,
        startZoneOffset: iv.start.offset,
        endTime: iv.end.instant,
        endZoneOffset: iv.end.offset,
        milliliters: milliliters,
      ),
      sourceTimeRange: AppleImportTimeRange(iv.start.instant, iv.end.instant),
      unit: record.unit,
      value: record.valueForReport,
    );
  }

  ConvertedAppleRecord? convertOxygenSaturation(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final value = record.numericValue;
    final percent = value == null ? null : toPercentage(value, record.unit);
    if (percent == null || percent < 0.0 || percent > 100.0) {
      return invalidRecord(
        record,
        'Oxygen saturation is missing, unsupported unit, or outside 0..100%.',
      );
    }
    return _instantRecord(
      record,
      start,
      'OxygenSaturationRecord',
      'spo2',
      (clientRecordId) => OxygenSaturationImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        percent: percent,
      ),
    );
  }

  ConvertedAppleRecord? convertRespiratoryRate(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final rate = record.numericValue;
    if (rate == null || rate <= 0.0) {
      return invalidRecord(
        record,
        'Respiratory rate is missing or not positive.',
      );
    }
    return _instantRecord(
      record,
      start,
      'RespiratoryRateRecord',
      'respiratory_rate',
      (clientRecordId) => RespiratoryRateImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        rate: rate,
      ),
    );
  }

  ConvertedAppleRecord? convertBodyTemperature(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final value = record.numericValue;
    final celsius = value == null ? null : toCelsius(value, record.unit);
    if (celsius == null) {
      return invalidRecord(
        record,
        'Body temperature is missing or has an unsupported unit.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BodyTemperatureRecord',
      'body_temperature',
      (clientRecordId) => BodyTemperatureImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        celsius: celsius,
      ),
    );
  }

  ConvertedAppleRecord? convertBloodGlucose(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final value = record.numericValue;
    final glucose =
        value == null ? null : toMillimolesPerLiter(value, record.unit);
    if (glucose == null) {
      return invalidRecord(
        record,
        'Blood glucose is missing or has an unsupported unit.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BloodGlucoseRecord',
      'blood_glucose',
      (clientRecordId) => BloodGlucoseImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        millimolesPerLiter: glucose,
      ),
    );
  }

  ConvertedAppleRecord? convertVo2Max(AppleRecord record, AppleDateTime start) {
    final vo2 = record.numericValue;
    if (vo2 == null || vo2 < 1.0 || vo2 > 100.0) {
      return invalidRecord(
        record,
        'VO2 max is missing or outside 1..100 mL/kg/min.',
      );
    }
    return _instantRecord(
      record,
      start,
      'Vo2MaxRecord',
      'vo2_max',
      (clientRecordId) => Vo2MaxImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        vo2MillilitersPerMinuteKilogram: vo2,
      ),
    );
  }

  ConvertedAppleRecord? convertBasalBodyTemperature(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final value = record.numericValue;
    final celsius = value == null ? null : toCelsius(value, record.unit);
    if (celsius == null) {
      return invalidRecord(
        record,
        'Basal body temperature is missing or has an unsupported unit.',
      );
    }
    return _instantRecord(
      record,
      start,
      'BasalBodyTemperatureRecord',
      'basal_body_temperature',
      (clientRecordId) => BasalBodyTemperatureImportRecord(
        clientRecordId: clientRecordId,
        time: start.instant,
        zoneOffset: start.offset,
        celsius: celsius,
      ),
    );
  }

  ConvertedAppleRecord? convertMindfulness(
    AppleRecord record,
    AppleInterval iv,
  ) {
    if (!mindfulnessAvailable) {
      return skippedNull(
        record,
        'feature_unavailable',
        'Mindfulness sessions are not available in this Health Connect provider.',
      );
    }
    final fingerprint = record.stableClientRecordId('mindfulness');
    markConverted(record.type);
    return ConvertedAppleRecord(
      appleType: record.type,
      targetType: 'MindfulnessSessionRecord',
      fingerprint: fingerprint,
      record: MindfulnessSessionImportRecord(
        clientRecordId:
            _metadataClientRecordId(record, 'MindfulnessSessionRecord'),
        startTime: iv.start.instant,
        startZoneOffset: iv.start.offset,
        endTime: iv.end.instant,
        endZoneOffset: iv.end.offset,
        title: 'Apple Health mindfulness',
      ),
      sourceTimeRange: AppleImportTimeRange(iv.start.instant, iv.end.instant),
      unit: record.unit,
      value: record.valueForReport,
    );
  }

  ConvertedAppleRecord? convertCycleCategory(
    AppleRecord record,
    AppleDateTime start,
  ) {
    final fingerprint = record.stableClientRecordId('cycle');
    final clientRecordId =
        _metadataClientRecordId(record, _substringAfterLast(record.type, 'Identifier'));
    final rawValue = record.rawValue ?? '';
    final ImportRecord importRecord;
    switch (record.type) {
      case appleMenstrualFlow:
        importRecord = MenstruationFlowImportRecord(
          clientRecordId: clientRecordId,
          time: start.instant,
          zoneOffset: start.offset,
          flow: mapMenstrualFlow(rawValue),
        );
      case appleOvulationTest:
        importRecord = OvulationTestImportRecord(
          clientRecordId: clientRecordId,
          time: start.instant,
          zoneOffset: start.offset,
          result: mapOvulationResult(rawValue),
        );
      case appleCervicalMucus:
        importRecord = CervicalMucusImportRecord(
          clientRecordId: clientRecordId,
          time: start.instant,
          zoneOffset: start.offset,
          appearance: mapCervicalMucusAppearance(rawValue),
          sensation: CervicalMucusSensation.unknown,
        );
      case appleIntermenstrualBleeding:
        importRecord = IntermenstrualBleedingImportRecord(
          clientRecordId: clientRecordId,
          time: start.instant,
          zoneOffset: start.offset,
        );
      case appleSexualActivity:
        importRecord = SexualActivityImportRecord(
          clientRecordId: clientRecordId,
          time: start.instant,
          zoneOffset: start.offset,
          protectionUsed: mapProtectionUsed(record.metadata),
        );
      default:
        return unsupportedNull(
          record,
          'No direct cycle mapping is implemented for this Apple record type.',
        );
    }
    markConverted(record.type);
    return ConvertedAppleRecord(
      appleType: record.type,
      targetType: importRecord.targetType,
      fingerprint: fingerprint,
      record: importRecord,
      sourceTimeRange: AppleImportTimeRange(start.instant, start.instant),
      unit: record.unit,
      value: record.valueForReport,
    );
  }
}

File: app/src/main/kotlin/tech/mmarca/openvitals/features/imports/applehealth/AppleHealthImportRecordConversions.kt
package tech.mmarca.openvitals.features.imports.applehealth

import androidx.health.connect.client.records.BasalBodyTemperatureRecord
import androidx.health.connect.client.records.BloodGlucoseRecord
import androidx.health.connect.client.records.BodyFatRecord
import androidx.health.connect.client.records.BodyTemperatureRecord
import androidx.health.connect.client.records.BodyWaterMassRecord
import androidx.health.connect.client.records.BoneMassRecord
import androidx.health.connect.client.records.CervicalMucusRecord
import androidx.health.connect.client.records.HeightRecord
import androidx.health.connect.client.records.HydrationRecord
import androidx.health.connect.client.records.IntermenstrualBleedingRecord
import androidx.health.connect.client.records.LeanBodyMassRecord
import androidx.health.connect.client.records.MealType
import androidx.health.connect.client.records.MenstruationFlowRecord
import androidx.health.connect.client.records.MindfulnessSessionRecord
import androidx.health.connect.client.records.OvulationTestRecord
import androidx.health.connect.client.records.OxygenSaturationRecord
import androidx.health.connect.client.records.Record
import androidx.health.connect.client.records.RespiratoryRateRecord
import androidx.health.connect.client.records.SexualActivityRecord
import androidx.health.connect.client.records.Vo2MaxRecord
import androidx.health.connect.client.records.WeightRecord
import androidx.health.connect.client.records.metadata.Metadata
import androidx.health.connect.client.units.Volume
import androidx.health.connect.client.units.celsius
import androidx.health.connect.client.units.kilograms
import androidx.health.connect.client.units.meters
import androidx.health.connect.client.units.percent

internal fun AppleHealthImportConverter.convertWeight(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val kg = record.numericValue?.toKilograms(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Weight is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("weight")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "WeightRecord",
        fingerprint,
        WeightRecord::class,
        WeightRecord(start.instant, start.offset, kg.kilograms, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertHeight(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val meters = record.numericValue?.toMeters(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Height is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("height")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "HeightRecord",
        fingerprint,
        HeightRecord::class,
        HeightRecord(start.instant, start.offset, meters.meters, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBodyFat(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val percent = record.numericValue?.toPercentage(record.unit)?.takeIf { it in 0.0..100.0 }
        ?: return invalid(record, "Body fat is missing, unsupported unit, or outside 0..100%.")
    val fingerprint = record.stableClientRecordId("body_fat")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BodyFatRecord",
        fingerprint,
        BodyFatRecord::class,
        BodyFatRecord(start.instant, start.offset, percent.percent, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertLeanMass(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val kg = record.numericValue?.toKilograms(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Lean body mass is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("lean_mass")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "LeanBodyMassRecord",
        fingerprint,
        LeanBodyMassRecord::class,
        LeanBodyMassRecord(start.instant, start.offset, kg.kilograms, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBoneMass(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val kg = record.numericValue?.toKilograms(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Bone mass is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("bone_mass")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BoneMassRecord",
        fingerprint,
        BoneMassRecord::class,
        BoneMassRecord(start.instant, start.offset, kg.kilograms, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBodyWaterMass(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val kg = record.numericValue?.toKilograms(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Body water mass is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("body_water_mass")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BodyWaterMassRecord",
        fingerprint,
        BodyWaterMassRecord::class,
        BodyWaterMassRecord(start.instant, start.offset, kg.kilograms, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertHydration(
    record: AppleRecord,
    interval: AppleInterval,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val milliliters = record.numericValue?.toMilliliters(record.unit)?.takeIf { it > 0.0 }
        ?: return invalid(record, "Hydration is missing, unsupported unit, or not positive.")
    val fingerprint = record.stableClientRecordId("hydration")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "HydrationRecord",
        fingerprint,
        HydrationRecord::class,
        HydrationRecord(
            startTime = interval.start.instant,
            startZoneOffset = interval.start.offset,
            endTime = interval.end.instant,
            endZoneOffset = interval.end.offset,
            volume = Volume.milliliters(milliliters),
            metadata = metadata,
        ),
        AppleImportTimeRange(interval.start.instant, interval.end.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertOxygenSaturation(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val percent = record.numericValue?.toPercentage(record.unit)?.takeIf { it in 0.0..100.0 }
        ?: return invalid(record, "Oxygen saturation is missing, unsupported unit, or outside 0..100%.")
    val fingerprint = record.stableClientRecordId("spo2")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "OxygenSaturationRecord",
        fingerprint,
        OxygenSaturationRecord::class,
        OxygenSaturationRecord(start.instant, start.offset, percent.percent, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertRespiratoryRate(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val rate = record.numericValue?.takeIf { it > 0.0 }
        ?: return invalid(record, "Respiratory rate is missing or not positive.")
    val fingerprint = record.stableClientRecordId("respiratory_rate")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "RespiratoryRateRecord",
        fingerprint,
        RespiratoryRateRecord::class,
        RespiratoryRateRecord(start.instant, start.offset, rate, metadata),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBodyTemperature(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val celsius = record.numericValue?.toCelsius(record.unit)
        ?: return invalid(record, "Body temperature is missing or has an unsupported unit.")
    val fingerprint = record.stableClientRecordId("body_temperature")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BodyTemperatureRecord",
        fingerprint,
        BodyTemperatureRecord::class,
        BodyTemperatureRecord(
            time = start.instant,
            zoneOffset = start.offset,
            temperature = celsius.celsius,
            metadata = metadata,
        ),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBloodGlucose(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val glucose = record.numericValue?.toBloodGlucose(record.unit)
        ?: return invalid(record, "Blood glucose is missing or has an unsupported unit.")
    val fingerprint = record.stableClientRecordId("blood_glucose")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BloodGlucoseRecord",
        fingerprint,
        BloodGlucoseRecord::class,
        BloodGlucoseRecord(
            time = start.instant,
            zoneOffset = start.offset,
            metadata = metadata,
            level = glucose,
            specimenSource = BloodGlucoseRecord.SPECIMEN_SOURCE_UNKNOWN,
            mealType = MealType.MEAL_TYPE_UNKNOWN,
            relationToMeal = BloodGlucoseRecord.RELATION_TO_MEAL_UNKNOWN,
        ),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertVo2Max(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val vo2 = record.numericValue?.takeIf { it in 1.0..100.0 }
        ?: return invalid(record, "VO2 max is missing or outside 1..100 mL/kg/min.")
    val fingerprint = record.stableClientRecordId("vo2_max")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "Vo2MaxRecord",
        fingerprint,
        Vo2MaxRecord::class,
        Vo2MaxRecord(
            time = start.instant,
            zoneOffset = start.offset,
            metadata = metadata,
            vo2MillilitersPerMinuteKilogram = vo2,
            measurementMethod = Vo2MaxRecord.MEASUREMENT_METHOD_OTHER,
        ),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertBasalBodyTemperature(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val celsius = record.numericValue?.toCelsius(record.unit)
        ?: return invalid(record, "Basal body temperature is missing or has an unsupported unit.")
    val fingerprint = record.stableClientRecordId("basal_body_temperature")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "BasalBodyTemperatureRecord",
        fingerprint,
        BasalBodyTemperatureRecord::class,
        BasalBodyTemperatureRecord(
            time = start.instant,
            zoneOffset = start.offset,
            metadata = metadata,
            temperature = celsius.celsius,
        ),
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertMindfulness(
    record: AppleRecord,
    interval: AppleInterval,
    metadata: Metadata,
): ConvertedAppleRecord? {
    if (!mindfulnessAvailable) {
        return skippedNull(record, "feature_unavailable", "Mindfulness sessions are not available in this Health Connect provider.")
    }
    val fingerprint = record.stableClientRecordId("mindfulness")
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        "MindfulnessSessionRecord",
        fingerprint,
        MindfulnessSessionRecord::class,
        MindfulnessSessionRecord(
            startTime = interval.start.instant,
            startZoneOffset = interval.start.offset,
            endTime = interval.end.instant,
            endZoneOffset = interval.end.offset,
            metadata = metadata,
            mindfulnessSessionType = MindfulnessSessionRecord.MINDFULNESS_SESSION_TYPE_UNKNOWN,
            title = "Apple Health mindfulness",
        ),
        AppleImportTimeRange(interval.start.instant, interval.end.instant),
        record.unit,
        record.valueForReport,
    )
}

internal fun AppleHealthImportConverter.convertCycleCategory(
    record: AppleRecord,
    start: AppleDateTime,
    metadata: Metadata,
): ConvertedAppleRecord? {
    val fingerprint = record.stableClientRecordId("cycle")
    val rawValue = record.rawValue.orEmpty()
    val convertedRecord: Record =
        when (record.type) {
            AppleMenstrualFlow -> MenstruationFlowRecord(
                time = start.instant,
                zoneOffset = start.offset,
                metadata = metadata,
                flow = rawValue.toMenstrualFlow(),
            )
            AppleOvulationTest -> OvulationTestRecord(
                time = start.instant,
                zoneOffset = start.offset,
                result = rawValue.toOvulationResult(),
                metadata = metadata,
            )
            AppleCervicalMucus -> CervicalMucusRecord(
                time = start.instant,
                zoneOffset = start.offset,
                metadata = metadata,
                appearance = rawValue.toCervicalMucusAppearance(),
                sensation = CervicalMucusRecord.SENSATION_UNKNOWN,
            )
            AppleIntermenstrualBleeding -> IntermenstrualBleedingRecord(
                time = start.instant,
                zoneOffset = start.offset,
                metadata = metadata,
            )
            AppleSexualActivity -> SexualActivityRecord(
                time = start.instant,
                zoneOffset = start.offset,
                metadata = metadata,
                protectionUsed = record.metadata.toProtectionUsed(),
            )
            else -> return unsupportedNull(record, "No direct cycle mapping is implemented for this Apple record type.")
        }
    markConverted(record.type)
    return ConvertedAppleRecord(
        record.type,
        convertedRecord::class.simpleName ?: "Record",
        fingerprint,
        convertedRecord::class,
        convertedRecord,
        AppleImportTimeRange(start.instant, start.instant),
        record.unit,
        record.valueForReport,
    )
}
